File: dotnet/Algoritmos/Searching/BinarySearch.cs
namespace Algoritmos.Searching;

/// <summary>
/// Binary search is an efficient algorithm for finding a specific value
/// in a sorted collection (e.g., an array). It works by repeatedly dividing
/// the search interval in half until the target value is found or
/// the interval becomes empty.
/// </summary>
public static class BinarySearch
{
    /// <summary>
    /// Steps:
    /// 1- Initialize: Set the start and end pointers to the beginning and end of the array.
    /// 2- Midpoint: Calculate the midpoint of the current interval.
    /// 3- Compare: Compare the value at the midpoint with the target value.
    /// 3.1- If they are equal, the search is successful, and the index of the midpoint is returned.
    /// 3.2- If the target value is less than the value at the midpoint, update the end pointer to be one less than the midpoint.
    /// 3.3- If the target value is greater than the value at the midpoint, update the start pointer to be one more than the midpoint.
    /// </summary>
    /// <param name="inputs">Array of SORTED inputs.</param>
    /// <param name="target">The value we want to find.</param>
    /// <returns>Index where the target is found in the array.</returns>
    public static int Find(int[] inputs, int target)
    {
        // Step 1: Set the start and end pointers
        int start = 0;
        int end = inputs.Length - 1;

        while (start <= end)
        {
            // Step 2: Calculate the midpoint of the current interval
            int mid = (start + end) / 2;
            
            // 3.1: Check if target is present at mid
            if (inputs[mid] == target)
            {
                return mid;
            }
                
            // 3.2: If target is greater, ignore the left half
            if (inputs[mid] < target)
            {
                start = mid + 1;
            }
            else
            {
                // 3.3: If target is smaller, ignore the right half
                end = mid - 1;
            }
        }
        
        return 0;
    }
}

File: typescript/src/searching/binary-search.ts
export function algorithmForBinarySearch(sortedInput: number[], target: number): number {
    // Step 1: Start with the entire array.
    // Define two pointers: left (first index) and right (last index)
    let left = 0;
    let right = sortedInput.length - 1;

    while (left <= right) {
        // Step 2: Find the middle index
        const middle = Math.floor((left + right) / 2);

        // Step 3: Check if the middle element is equal to the target.
        // If it is, return the index of the middle element.
        if (sortedInput[middle] === target) {
            return middle;
        }

        // If it’s less than the target, discard the left half by moving left to middle + 1.
        if (sortedInput[middle] < target) {
            left = middle + 1;
        } else {
            // If it’s greater than the target, discard the right half by moving right to middle - 1.
            right = middle - 1;
        }
    }

    // Target not found
    return -1;
}
/*
Complexity
    Time Complexity: O(log⁡n)O(logn), as the array is divided in half with each step.
    Space Complexity: O(1)O(1), as it only uses a constant amount of extra space.

Pros

    Efficiency:
    Binary search is much faster than a linear search for large datasets.
    Its time complexity is O(log⁡n)O(logn), so it scales well as the size of the array grows.

    Low Memory Usage:
    The iterative version of binary search only uses a constant amount of extra space (O(1)O(1)),
    making it memory efficient.

    Simple to Implement:
    Despite its efficiency, binary search is relatively simple to implement once the logic
    is understood.

    Predictable Performance:
    Since binary search always divides the search range in half, its performance is predictable
    and stable.

    Widely Applicable in Sorted Data:
    Binary search is highly useful in applications involving sorted data, such as finding
    elements in databases, searching in dictionaries, or looking up values in maps.

Cons

    Requires Sorted Data:
    Binary search only works on sorted arrays. If the data isn’t already sorted, you’ll
    need to sort it first, which could increase complexity, especially for large datasets.

    Not Optimal for Small Data:
    For small datasets, linear search can sometimes be more efficient since binary search
    requires additional overhead to calculate the middle index and compare values.

    Inflexible for Dynamic Data:
    Binary search is less efficient for dynamically changing data. Every time the array is
    modified (insertion or deletion), it needs to be re-sorted, which can be costly.

    Reduced Performance with Linked Lists:
    Binary search is not well-suited for linked lists because accessing the middle element
    in a linked list requires traversing the list, which makes the performance O(n)O(n).

    Recursive Implementation Risk:
    In some languages, a recursive implementation of binary search can lead to stack overflow
    for very large arrays due to deep recursion levels. This is avoided with an iterative
    approach but is a consideration in recursive cases.

When to Use Binary Search

    Best used on large, sorted arrays where search speed is critical.
    Avoid if the data isn’t sorted, if you frequently modify the data, or if you’re working
    with data structures that don’t support fast indexing (e.g., linked lists).
 */